src/graphene.cpp

```diff
@@ -0,0 +1,1504 @@
+    uint64_t nOriginalSizeTotal = 0;
+    for (map<int64_t, pair<uint64_t, uint64_t> >::iterator mi = mapGrapheneBlocksOutBound.begin();
+         mi != mapGrapheneBlocksOutBound.end(); ++mi)
+    {
+        nGrapheneSizeTotal += (*mi).second.first;
+        nOriginalSizeTotal += (*mi).second.second;
+    }
+    // We count up the inbound CMemPoolInfo sizes. Inbound CMemPoolInfo sizes go with Outbound graphene blocks.
+    uint64_t nInBoundMemPoolInfoSize = 0;
+    for (map<int64_t, uint64_t>::iterator mi = mapMemPoolInfoInBound.begin(); mi != mapMemPoolInfoInBound.end(); ++mi)
+        nInBoundMemPoolInfoSize += (*mi).second;
+
+    if (nOriginalSizeTotal > 0)
+        nCompressionRate = 100 - (100 * (double)(nGrapheneSizeTotal + nInBoundMemPoolInfoSize) / nOriginalSizeTotal);
+
+    ostringstream ss;
+    ss << fixed << setprecision(1);
+    ss << "Compression for " << mapGrapheneBlocksOutBound.size()
+       << " Outbound graphene blocks (last 24hrs): " << nCompressionRate << "%";
+    return ss.str();
+}
+
+// Calculate the average inbound graphene CMemPoolInfo size
+string CGrapheneBlockData::InBoundMemPoolInfoToString()
+{
+    LOCK(cs_graphenestats);
+    double avgMemPoolInfoSize = average(mapMemPoolInfoInBound);
+    ostringstream ss;
+    ss << "Inbound CMemPoolInfo size (last 24hrs) AVG: " << formatInfoUnit(avgMemPoolInfoSize);
+    return ss.str();
+}
+
+// Calculate the average inbound graphene CMemPoolInfo size
+string CGrapheneBlockData::OutBoundMemPoolInfoToString()
+{
+    LOCK(cs_graphenestats);
+    double avgMemPoolInfoSize = average(mapMemPoolInfoOutBound);
+    ostringstream ss;
+    ss << "Outbound CMemPoolInfo size (last 24hrs) AVG: " << formatInfoUnit(avgMemPoolInfoSize);
+    return ss.str();
+}
+
+// Calculate the graphene percentage compression over the last 24 hours
+string CGrapheneBlockData::ResponseTimeToString()
+{
+    LOCK(cs_graphenestats);
+
+    vector<double> vResponseTime;
+
+    double nResponseTimeAverage = 0;
+    double nPercentile = 0;
+    double nTotalResponseTime = 0;
+    double nTotalEntries = 0;
+    for (map<int64_t, double>::iterator mi = mapGrapheneBlockResponseTime.begin();
+         mi != mapGrapheneBlockResponseTime.end(); ++mi)
+    {
+        nTotalEntries += 1;
+        nTotalResponseTime += (*mi).second;
+        vResponseTime.push_back((*mi).second);
+    }
+
+    if (nTotalEntries > 0)
+    {
+        nResponseTimeAverage = (double)nTotalResponseTime / nTotalEntries;
+
+        // Calculate the 95th percentile
+        uint64_t nPercentileElement = static_cast<int>((nTotalEntries * 0.95) + 0.5) - 1;
+        sort(vResponseTime.begin(), vResponseTime.end());
+        nPercentile = vResponseTime[nPercentileElement];
+    }
+
+    ostringstream ss;
+    ss << fixed << setprecision(2);
+    ss << "Response time   (last 24hrs) AVG:" << nResponseTimeAverage << ", 95th pcntl:" << nPercentile;
+    return ss.str();
+}
+
+// Calculate the graphene percentage compression over the last 24 hours
```

Comment:
stale comment. this function compute the avg block validation time in the last 24 hrs

Reply:
Done.